ddl/reorg.go

```diff
@@ -479,32 +501,59 @@ func getReorgInfo(d *ddlCtx, t *meta.Meta, job *model.Job, tbl table.Table) (*re
 		failpoint.Inject("errorUpdateReorgHandle", func() (*reorgInfo, error) {
 			return &info, errors.New("occur an error when update reorg handle")
 		})
-		err = t.UpdateDDLReorgHandle(job, start, end, pid)
+		err = t.UpdateDDLReorgHandle(job, start, end, pid, elements[0])
 		if err != nil {
 			return &info, errors.Trace(err)
 		}
 		// Update info should after data persistent.
 		job.SnapshotVer = ver.Ver
+		element = elements[0]
 	} else {
+		failpoint.Inject("MockGetIndexRecordErr", func(val failpoint.Value) {
+			// For the case of the old TiDB version(do not exist the element information) is upgraded to the new TiDB version.
+			// Second step, we need to remove the element information to make sure we can get the error of "ErrDDLReorgElementNotExist".
+			// However, since "txn.Reset()" will be called later, the reorganized information cannot be saved to storage.
+			if val.(string) == "addIdxNotOwnerErr" && atomic.CompareAndSwapUint32(&mockNotOwnerErrOnce, 2, 3) {
+				if err := t.RemoveReorgElement(job); err != nil {
+					failpoint.Return(nil, errors.Trace(err))
+				}
+			}
+		})
+
 		var err error
-		start, end, pid, err = t.GetDDLReorgHandle(job, tbl.Meta().IsCommonHandle)
+		element, start, end, pid, err = t.GetDDLReorgHandle(job, tbl.Meta().IsCommonHandle)
 		if err != nil {
-			return nil, errors.Trace(err)
+			// If the reorg element doesn't exist, this reorg info should be saved by the older TiDB versions.
+			// It's compatible with the older TiDB versions.
+			if meta.ErrDDLReorgElementNotExist.Equal(err) {
```

Comment:
What about add a comment that this logic can be removed in next TiDB version?

Reply:
Done.